wp/plugin/lib/utils.ml

```diff
@@ -54,26 +54,29 @@ let clear_mapper : Term.mapper = object
     super#map_term cls t'
 end
 
-(* Reads in the program_t from a file. *)
+(* Reads in the program_t and its architecture from a file. *)
 let read_program (ctxt : ctxt) ~(loader : string) ~(filepath : string)
-    : Program.t =
+    : Program.t * Arch.t =
   let mk_digest = Cache.Digests.get_generator ctxt ~filepath ~loader in
   let knowledge_digest = Cache.Digests.knowledge mk_digest in
   let () = Cache.Knowledge.load knowledge_digest in
   match Cache.Program.load filepath with
   | Some prog ->
     info "Program %s (%a) found in cache.%!"
       filepath Data.Cache.Digest.pp knowledge_digest;
-    prog
+    prog, Option.value_exn (Cache.Arch.load filepath)
   | None ->
     (* The program_t is not in the cache. Disassemble the binary. *)
     info "Saving program %s (%a) to cache.%!"
       filepath Data.Cache.Digest.pp knowledge_digest;
     let project = create_proj None loader filepath in
     let prog = project |> Project.program |> clear_mapper#run in
+    let arch = Project.arch project in
+    let () = Toplevel.reset () in
```

Comment:
This will break the inner consistency of the knowledge base. You have created a project, which is a value that depends on the knowledge base state, then you reset the state and store only the project, therefore all the objects that the projet references disappear. Even worse, any consequent computations will create new objects without knowing that they were already created in the previous knowledge base and thus unwated aliasing will be created. In other words, it is totally undefined behavior. 

If you just want to save the `Program.t` what stops you from storing it directly? Why put it in the knowledge base?

Reply:
This is true. I can definitely create a cache for just the `Program.t`. This way, I wouldn't have to mess with the inner consistency of the KB. Thanks!

Reply:
I would even suggest just storing in a file without using cache. So that you can disable cache and work only with Program.t or even with Sub.t to enable the granularity that you need.

Reply:
I have two ideas:
1. Remove our use of the knowledge and state caches. Replace it with caches for `Program.t` and `Arch.t`. If its possible, combine those into the same cache.
2. We bring back our `save_project` plugin, but we serialize both `Program.t` and `Arch.t` information to our `.bpj` file. We create a flag that indicates that the file we are reading is a `.bpj` file rather than the binary itself.

I'm more inclined to the first idea.

Reply:
I also have a mild preference for the first idea, primarily because it keeps the CLI clean.

Reply:
You can also specify your own loader that will read your stored data into the `Projec.t` representation, in #249 I even suggested to store one subroutine per file to minimize the bap startup overhead so that parallelizing wp could be done easily.